File: parrot/inserter.py
from __future__ import annotations

import time

import pyautogui
import pyperclip


class TextInserter:
    def __init__(self, restore_clipboard: bool, paste_delay_seconds: float) -> None:
        self.restore_clipboard = restore_clipboard
        self.paste_delay_seconds = paste_delay_seconds

    def paste(self, text: str) -> None:
        if not text.strip():
            return

        old_clipboard = None
        if self.restore_clipboard:
            try:
                old_clipboard = pyperclip.paste()
            except pyperclip.PyperclipException:
                old_clipboard = None

        pyperclip.copy(text)
        time.sleep(self.paste_delay_seconds)
        pyautogui.hotkey("ctrl", "v")

        if self.restore_clipboard and old_clipboard is not None:
            time.sleep(self.paste_delay_seconds)
            pyperclip.copy(old_clipboard)

File: src/inserter.rs
use std::{thread, time::Duration};

use anyhow::{Context, Result};
use arboard::Clipboard;
use rdev::{simulate, EventType, Key};

#[derive(Clone)]
pub struct TextInserter {
    restore_clipboard: bool,
}

impl TextInserter {
    pub fn new(restore_clipboard: bool) -> Self {
        Self { restore_clipboard }
    }

    pub fn paste(&self, text: &str) -> Result<()> {
        let mut clipboard = Clipboard::new().context("failed to open clipboard")?;
        let old_clipboard = if self.restore_clipboard {
            clipboard.get_text().ok()
        } else {
            None
        };

        clipboard
            .set_text(text.to_string())
            .context("failed to set clipboard text")?;
        thread::sleep(Duration::from_millis(80));
        send_ctrl_v()?;

        if let Some(old_clipboard) = old_clipboard {
            thread::sleep(Duration::from_millis(80));
            let _ = clipboard.set_text(old_clipboard);
        }
        Ok(())
    }
}

fn send_ctrl_v() -> Result<()> {
    simulate(&EventType::KeyPress(Key::ControlLeft))
        .map_err(|_| anyhow::anyhow!("failed to press Ctrl"))?;
    simulate(&EventType::KeyPress(Key::KeyV)).map_err(|_| anyhow::anyhow!("failed to press V"))?;
    simulate(&EventType::KeyRelease(Key::KeyV))
        .map_err(|_| anyhow::anyhow!("failed to release V"))?;
    simulate(&EventType::KeyRelease(Key::ControlLeft))
        .map_err(|_| anyhow::anyhow!("failed to release Ctrl"))?;
    Ok(())
}
